// @glint/core/lib/language-server/util/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeFilePath = exports.filePathToUri = exports.uriToFilePath = exports.scriptElementKindToCompletionItemKind = exports.offsetToPosition = exports.positionToOffset = void 0;
var position_1 = require("./position");
Object.defineProperty(exports, "positionToOffset", { enumerable: true, get: function () { return position_1.positionToOffset; } });
Object.defineProperty(exports, "offsetToPosition", { enumerable: true, get: function () { return position_1.offsetToPosition; } });
var protocol_1 = require("./protocol");
Object.defineProperty(exports, "scriptElementKindToCompletionItemKind", { enumerable: true, get: function () { return protocol_1.scriptElementKindToCompletionItemKind; } });
const vscode_uri_1 = require("vscode-uri");
function uriToFilePath(uri) {
    return vscode_uri_1.URI.parse(uri).fsPath.replace(/\\/g, '/');
}
exports.uriToFilePath = uriToFilePath;
function filePathToUri(filePath) {
    return vscode_uri_1.URI.file(filePath).toString();
}
exports.filePathToUri = filePathToUri;
function normalizeFilePath(filePath) {
    return uriToFilePath(filePathToUri(filePath));
}
exports.normalizeFilePath = normalizeFilePath;
