// @glint/core/lib/cli/perform-check.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.performCheck = void 0;
const transform_manager_1 = __importDefault(require("../common/transform-manager"));
const diagnostics_1 = require("./diagnostics");
function performCheck(glintConfig, optionsToExtend) {
    let { ts } = glintConfig;
    let transformManager = new transform_manager_1.default(glintConfig);
    let parsedConfig = loadTsconfig(ts, transformManager, glintConfig.configPath, optionsToExtend);
    let compilerHost = createCompilerHost(ts, parsedConfig.options, transformManager);
    let formatDiagnostic = (0, diagnostics_1.buildDiagnosticFormatter)(ts);
    let program = ts.createProgram({
        rootNames: parsedConfig.fileNames,
        options: parsedConfig.options,
        host: compilerHost,
    });
    program.emit();
    let baselineDiagnostics = collectDiagnostics(program, transformManager, parsedConfig.options);
    let fullDiagnostics = transformManager.rewriteDiagnostics(baselineDiagnostics);
    for (let diagnostic of fullDiagnostics) {
        console.error(formatDiagnostic(diagnostic));
    }
    process.exit(fullDiagnostics.length ? 1 : 0);
}
exports.performCheck = performCheck;
function collectDiagnostics(program, transformManager, options) {
    return [
        ...program.getSyntacticDiagnostics(),
        ...transformManager.getTransformDiagnostics(),
        ...program.getSemanticDiagnostics(),
        ...(options.declaration ? program.getDeclarationDiagnostics() : []),
    ];
}
function createCompilerHost(ts, options, transformManager) {
    let host = ts.createCompilerHost(options);
    host.fileExists = transformManager.fileExists;
    host.readFile = transformManager.readTransformedFile;
    host.readDirectory = transformManager.readDirectory;
    return host;
}
function loadTsconfig(ts, transformManager, configPath, optionsToExtend) {
    if (!configPath) {
        return {
            fileNames: [],
            options: optionsToExtend,
            errors: [],
        };
    }
    let config = ts.getParsedCommandLineOfConfigFile(configPath, optionsToExtend, {
        ...ts.sys,
        readDirectory: transformManager.readDirectory,
        onUnRecoverableConfigFileDiagnostic(diagnostic) {
            let { messageText } = diagnostic;
            if (typeof messageText !== 'string') {
                messageText = messageText.messageText;
            }
            throw new Error(messageText);
        },
    });
    if (!config) {
        throw new Error('Unknown error loading config');
    }
    return config;
}
