// @glint/core/lib/cli/perform-watch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.performWatch = void 0;
const transform_manager_1 = __importDefault(require("../common/transform-manager"));
const diagnostics_1 = require("./diagnostics");
function performWatch(glintConfig, optionsToExtend) {
    let { ts } = glintConfig;
    let transformManager = new transform_manager_1.default(glintConfig);
    let formatDiagnostic = (0, diagnostics_1.buildDiagnosticFormatter)(ts);
    let host = ts.createWatchCompilerHost(glintConfig.configPath, optionsToExtend, sysForWatchCompilerHost(ts, transformManager), ts.createSemanticDiagnosticsBuilderProgram, (diagnostic) => console.error(formatDiagnostic(diagnostic)));
    patchWatchCompilerHost(host, transformManager);
    ts.createWatchProgram(host);
}
exports.performWatch = performWatch;
function sysForWatchCompilerHost(ts, transformManager) {
    return {
        ...ts.sys,
        readDirectory: transformManager.readDirectory,
        watchDirectory: transformManager.watchDirectory,
        fileExists: transformManager.fileExists,
        watchFile: transformManager.watchTransformedFile,
        readFile: transformManager.readTransformedFile,
    };
}
function patchWatchCompilerHost(host, transformManager) {
    let { afterProgramCreate } = host;
    host.afterProgramCreate = (program) => {
        patchProgram(program, transformManager);
        afterProgramCreate === null || afterProgramCreate === void 0 ? void 0 : afterProgramCreate.call(host, program);
    };
}
function patchProgram(program, transformManager) {
    let { getSyntacticDiagnostics, getSemanticDiagnostics } = program;
    program.getSyntacticDiagnostics = function (sourceFile, cancelationToken) {
        let diagnostics = getSyntacticDiagnostics.call(program, sourceFile, cancelationToken);
        let transformDiagnostics = transformManager.getTransformDiagnostics(sourceFile === null || sourceFile === void 0 ? void 0 : sourceFile.fileName);
        return [...diagnostics, ...transformDiagnostics];
    };
    program.getSemanticDiagnostics = (sourceFile, cancellationToken) => {
        let diagnostics = getSemanticDiagnostics.call(program, sourceFile, cancellationToken);
        return transformManager.rewriteDiagnostics(diagnostics, sourceFile === null || sourceFile === void 0 ? void 0 : sourceFile.fileName);
    };
}
